File: python/src/solana_mpp/client/session.py
"""Client-side session intent implementation.

Tracks an open payment channel and signs cumulative vouchers for each API
call. Vouchers are Ed25519-signed over the on-chain Borsh voucher layout used
by the payment-channels program.

Mirrors the Rust spine at ``rust/crates/mpp/src/client/session.rs``.

The ``signer`` is any object exposing ``pubkey()`` and ``sign_message(bytes)``
that returns a 64-byte Ed25519 signature (e.g. ``solders.keypair.Keypair``).
"""

from __future__ import annotations

from typing import Any

from solana_mpp.protocol.session import (
    DEFAULT_SESSION_EXPIRES_AT,
    ClosePayload,
    OpenPayload,
    SessionActionPayload,
    SignedVoucher,
    TopUpPayload,
    VoucherData,
    VoucherPayload,
    session_action_to_dict,
)

# Default voucher expiry: 2100-01-01T00:00:00Z. Below JavaScript's max safe
# integer so JSON intermediaries do not round it before decoding.
DEFAULT_VOUCHER_EXPIRES_AT = DEFAULT_SESSION_EXPIRES_AT

# Push / pull session submission modes (re-exported for ergonomics).
SESSION_MODE_PUSH = "push"
SESSION_MODE_PULL = "pull"


def _signer_pubkey_b58(signer: Any) -> str:
    return str(signer.pubkey())


def _signature_b58(signer: Any, message: bytes) -> str:
    """Sign ``message`` and return the base58 signature.

    ``solders`` signers return a ``Signature`` whose ``str()`` is base58.
    """
    return str(signer.sign_message(message))


class ActiveSession:
    """Tracks the client-side state of an active payment session.

    Holds a signer (its public key becomes the ``authorizedSigner``) and
    advances the cumulative watermark with each signed voucher.
    """

    def __init__(self, channel_id: str, signer: Any, expires_at: int = DEFAULT_VOUCHER_EXPIRES_AT) -> None:
        self.channel_id = channel_id
        self.cumulative = 0
        self._nonce = 0
        self._expires_at = expires_at
        self._signer = signer

    def set_expires_at(self, expires_at: int) -> None:
        """Update the expiry used for subsequent vouchers."""
        self._expires_at = expires_at

    def authorized_signer(self) -> str:
        """The authorized signer public key (base58)."""
        return _signer_pubkey_b58(self._signer)

    def channel_id_str(self) -> str:
        """Channel ID as base58."""
        return self.channel_id

    def prepare_voucher(self, cumulative: int) -> SignedVoucher:
        """Prepare a signed voucher without advancing the local watermark.

        Useful for ack/commit transports: if sending the commit fails the
        client can retry the same cumulative without its state drifting ahead
        of the server.
        """
        if cumulative <= self.cumulative:
            raise ValueError(
                f"Voucher cumulative {cumulative} must exceed current watermark {self.cumulative}"
            )
        data = VoucherData(
            channel_id=self.channel_id_str(),
            cumulative=str(cumulative),
            expires_at=self._expires_at,
            nonce=self._nonce + 1,
        )
        signature = _signature_b58(self._signer, data.message_bytes())
        return SignedVoucher(data=data, signature=signature)

    def prepare_increment(self, amount: int) -> SignedVoucher:
        """Prepare a voucher adding ``amount`` without advancing the watermark."""
        return self.prepare_voucher(self.cumulative + amount)

    def record_voucher(self, voucher: SignedVoucher) -> None:
        """Record a prepared voucher as accepted by the server."""
        cumulative = int(voucher.data.cumulative)
        if cumulative <= self.cumulative:
            raise ValueError(
                f"Voucher cumulative {cumulative} must exceed current watermark {self.cumulative}"
            )
        self.cumulative = cumulative
        self._nonce = max(self._nonce, voucher.data.nonce if voucher.data.nonce is not None else self._nonce + 1)

    def sign_voucher(self, cumulative: int) -> SignedVoucher:
        """Sign a voucher with an absolute cumulative amount and advance state."""
        voucher = self.prepare_voucher(cumulative)
        self.record_voucher(voucher)
        return voucher

    def sign_increment(self, amount: int) -> SignedVoucher:
        """Sign a voucher adding ``amount`` to the current cumulative."""
        return self.sign_voucher(self.cumulative + amount)

    def voucher_action(self, amount: int) -> SessionActionPayload:
        """Build a ``voucher`` action wrapping a freshly-signed increment."""
        return VoucherPayload(voucher=self.sign_increment(amount))

    def close_action(self, final_increment: int | None = None) -> SessionActionPayload:
        """Build a ``close`` action for cooperative channel close.

        If ``final_increment`` is a positive int, signs one last voucher for
        the remaining balance before closing.
        """
        voucher: SignedVoucher | None = None
        if final_increment is not None and final_increment > 0:
            voucher = self.sign_increment(final_increment)
        return ClosePayload(channel_id=self.channel_id_str(), voucher=voucher)

    def open_action(self, deposit: int, open_tx_signature: str) -> SessionActionPayload:
        """Build an ``open`` action for push mode."""
        return OpenPayload.push(
            self.channel_id_str(),
            str(deposit),
            self.authorized_signer(),
            open_tx_signature,
        )

    def open_payment_channel_action(
        self,
        deposit: int,
        payer: str,
        payee: str,
        mint: str,
        salt: int,
        grace_period: int,
        open_tx_signature: str,
    ) -> SessionActionPayload:
        """Build an ``open`` action for the payment-channels program (push)."""
        return self.open_payment_channel_action_with_mode(
            SESSION_MODE_PUSH, deposit, payer, payee, mint, salt, grace_period, open_tx_signature
        )

    def open_payment_channel_action_with_mode(
        self,
        mode: str,
        deposit: int,
        payer: str,
        payee: str,
        mint: str,
        salt: int,
        grace_period: int,
        open_tx_signature: str,
    ) -> SessionActionPayload:
        """Build a payment-channel ``open`` action with an explicit mode."""
        return OpenPayload.payment_channel_with_mode(
            mode,
            self.channel_id_str(),
            str(deposit),
            payer,
            payee,
            mint,
            salt,
            grace_period,
            self.authorized_signer(),
            open_tx_signature,
        )

    def open_pull_action(self, approved_amount: int, owner: str, approve_tx_signature: str) -> SessionActionPayload:
        """Build an ``open`` action for pull mode (SPL token delegation)."""
        return OpenPayload.pull(
            self.channel_id_str(),  # token account used as the session identifier
            str(approved_amount),
            owner,
            self.authorized_signer(),
            approve_tx_signature,
        )

    def topup_action(self, new_deposit: int, topup_tx_signature: str) -> SessionActionPayload:
        """Build a ``topup`` action after a top-up transaction."""
        return TopUpPayload(
            channel_id=self.channel_id_str(),
            new_deposit=str(new_deposit),
            signature=topup_tx_signature,
        )


def action_to_dict(payload: SessionActionPayload) -> dict[str, Any]:
    """Serialize a session action payload as a tagged wire object."""
    return session_action_to_dict(payload)

File: swift/Sources/SolanaMpp/Client/Session.swift
import Foundation

// MARK: - Active session
//
// Client-side session intent implementation. Mirrors the Rust spine
// `rust/crates/mpp/src/client/session.rs` and
// `client/session_consumer.rs`: track an open payment channel and sign
// cumulative vouchers for each API call. Vouchers are Ed25519-signed over
// the on-chain Borsh voucher layout used by the payment-channels program.

/// Default voucher expiry: 2100-01-01T00:00:00Z. Stays below JavaScript's
/// max safe integer so JSON intermediaries do not round it.
public let DEFAULT_VOUCHER_EXPIRES_AT: Int64 = DEFAULT_SESSION_EXPIRES_AT

/// Tracks the client-side state of an active payment session.
///
/// Holds a `SolanaSigner` session key and advances the cumulative
/// watermark with each signed voucher. The signer may be a local memory
/// signer, a hardware wallet, or any cloud KMS via the `SolanaSigner`
/// protocol.
///
/// This is a reference type (`actor`-free `final class`) because the
/// watermark is read-modify-write state that must not be copied by value
/// when shared with a `SessionConsumer`.
public final class ActiveSession: @unchecked Sendable {
    /// On-chain channel address.
    public let channelId: Pubkey

    /// Cumulative amount authorized so far (base units).
    public private(set) var cumulative: UInt64 = 0

    /// Nonce counter, incremented with each signed voucher.
    private var nonce: UInt64 = 0

    /// Unix timestamp at which newly signed vouchers expire.
    private var expiresAt: Int64

    /// Session signing key.
    private let signer: any SolanaSigner

    public init(channelId: Pubkey, signer: any SolanaSigner) {
        self.channelId = channelId
        self.signer = signer
        self.expiresAt = DEFAULT_VOUCHER_EXPIRES_AT
    }

    public init(channelId: Pubkey, signer: any SolanaSigner, expiresAt: Int64) {
        self.channelId = channelId
        self.signer = signer
        self.expiresAt = expiresAt
    }

    /// Update the expiry timestamp used for subsequent vouchers.
    public func setExpiresAt(_ expiresAt: Int64) {
        self.expiresAt = expiresAt
    }

    /// The authorized signer public key (base58) for the `open` action.
    public func authorizedSigner() -> String {
        Base58.encode(signer.publicKey)
    }

    /// Channel ID as base58.
    public func channelIdString() -> String {
        channelId.base58
    }

    /// Sign a voucher with an absolute cumulative amount and advance the
    /// local watermark. `cumulative` MUST strictly exceed the current
    /// watermark.
    @discardableResult
    public func signVoucher(cumulative: UInt64) async throws -> SignedVoucher {
        let voucher = try await prepareVoucher(cumulative: cumulative)
        try recordVoucher(voucher)
        return voucher
    }

    /// Prepare a signed voucher without advancing the local watermark.
    ///
    /// Useful for ack/commit transports: if the commit fails, the client
    /// can retry the same cumulative amount without local state drifting
    /// ahead of the server.
    public func prepareVoucher(cumulative: UInt64) async throws -> SignedVoucher {
        guard cumulative > self.cumulative else {
            throw MppError.invalidTransaction(
                "Voucher cumulative \(cumulative) must exceed current watermark \(self.cumulative)"
            )
        }
        let data = VoucherData(
            channelId: channelIdString(),
            cumulative: String(cumulative),
            expiresAt: expiresAt,
            nonce: nonce + 1
        )
        let bytes = try data.messageBytes()
        let sig = try await signer.sign(message: bytes)
        guard sig.count == 64 else {
            throw MppError.signingFailure("signer returned \(sig.count) bytes, expected 64")
        }
        return SignedVoucher(data: data, signature: Base58.encode(sig))
    }

    /// Prepare a signed voucher adding `amount` without advancing the
    /// watermark.
    public func prepareIncrement(_ amount: UInt64) async throws -> SignedVoucher {
        try await prepareVoucher(cumulative: cumulative + amount)
    }

    /// Record a prepared voucher as accepted by the server.
    public func recordVoucher(_ voucher: SignedVoucher) throws {
        guard let cumulativeValue = UInt64(voucher.data.cumulative) else {
            throw MppError.invalidTransaction("invalid voucher cumulative")
        }
        guard cumulativeValue > cumulative else {
            throw MppError.invalidTransaction(
                "Voucher cumulative \(cumulativeValue) must exceed current watermark \(cumulative)"
            )
        }
        cumulative = cumulativeValue
        nonce = max(nonce, voucher.data.nonce ?? (nonce + 1))
    }

    /// Sign a voucher adding `amount` to the current cumulative.
    @discardableResult
    public func signIncrement(_ amount: UInt64) async throws -> SignedVoucher {
        try await signVoucher(cumulative: cumulative + amount)
    }

    /// Build a `SessionAction.voucher` wrapping a freshly-signed increment.
    public func voucherAction(_ amount: UInt64) async throws -> SessionAction {
        let voucher = try await signIncrement(amount)
        return .voucher(VoucherPayload(voucher: voucher))
    }

    /// Build a `SessionAction.close` for cooperative channel close.
    ///
    /// If `finalIncrement` is non-nil and > 0, signs one last voucher for
    /// the remaining balance before closing.
    public func closeAction(finalIncrement: UInt64?) async throws -> SessionAction {
        let voucher: SignedVoucher?
        if let amount = finalIncrement, amount > 0 {
            voucher = try await signIncrement(amount)
        } else {
            voucher = nil
        }
        return .close(ClosePayload(channelId: channelIdString(), voucher: voucher))
    }

    /// Build a `SessionAction.open` for **push** mode.
    public func openAction(deposit: UInt64, openTxSignature: String) -> SessionAction {
        .open(OpenPayload.push(
            channelId: channelIdString(),
            deposit: String(deposit),
            authorizedSigner: authorizedSigner(),
            signature: openTxSignature
        ))
    }

    /// Build a `SessionAction.open` for the payment-channels program.
    public func openPaymentChannelAction(
        deposit: UInt64,
        payer: String,
        payee: String,
        mint: String,
        salt: UInt64,
        gracePeriod: UInt32,
        openTxSignature: String
    ) -> SessionAction {
        openPaymentChannelAction(
            mode: .push,
            deposit: deposit,
            payer: payer,
            payee: payee,
            mint: mint,
            salt: salt,
            gracePeriod: gracePeriod,
            openTxSignature: openTxSignature
        )
    }

    /// Build a payment-channel `SessionAction.open` with an explicit mode.
    public func openPaymentChannelAction(
        mode: SessionMode,
        deposit: UInt64,
        payer: String,
        payee: String,
        mint: String,
        salt: UInt64,
        gracePeriod: UInt32,
        openTxSignature: String
    ) -> SessionAction {
        .open(OpenPayload.paymentChannel(
            mode: mode,
            channelId: channelIdString(),
            deposit: String(deposit),
            payer: payer,
            payee: payee,
            mint: mint,
            salt: salt,
            gracePeriod: gracePeriod,
            authorizedSigner: authorizedSigner(),
            signature: openTxSignature
        ))
    }

    /// Build a `SessionAction.open` for **pull** mode (SPL delegation).
    public func openPullAction(
        approvedAmount: UInt64,
        owner: String,
        approveTxSignature: String
    ) -> SessionAction {
        .open(OpenPayload.pull(
            tokenAccount: channelIdString(), // token account is the session id
            approvedAmount: String(approvedAmount),
            owner: owner,
            authorizedSigner: authorizedSigner(),
            signature: approveTxSignature
        ))
    }

    /// Build a `SessionAction.topUp` after a top-up transaction.
    public func topupAction(newDeposit: UInt64, topupTxSignature: String) -> SessionAction {
        .topUp(TopUpPayload(
            channelId: channelIdString(),
            newDeposit: String(newDeposit),
            signature: topupTxSignature
        ))
    }
}

// MARK: - Session dispatch entry point

/// High-level entry points for the Solana session client. Mirrors the
/// Rust spine `solana_mpp::client::session` dispatch surface: pick a
/// `solana` + `session` challenge from a multi-challenge 402 response,
/// decode the embedded `SessionRequest`, and frame a `SessionAction` into
/// an `Authorization: Payment ...` header value.
public enum Session {
    /// Returns the first `solana` + `session` challenge in a list of raw
    /// `WWW-Authenticate` header values whose embedded `SessionRequest`
    /// also decodes cleanly. Mirrors `Charge.pickChallenge`.
    public static func pickChallenge(wwwAuthenticateHeaders: [String]) throws -> PaymentChallenge {
        for header in wwwAuthenticateHeaders {
            guard let challenge = try? MppHeaders.parseWWWAuthenticate(header),
                  challenge.method == "solana", challenge.intent == "session" else {
                continue
            }
            guard (try? challenge.sessionRequest) != nil else { continue }
            return challenge
        }
        throw MppError.unsupportedChallenge(method: "(missing)", intent: "(missing)")
    }

    /// Frame a `SessionAction` into the `Authorization: Payment ...`
    /// header value, echoing the challenge the action responds to.
    public static func authorizationHeader(
        for challenge: PaymentChallenge,
        action: SessionAction
    ) throws -> String {
        try challenge.requireSolanaSession()
        let encoder = JSONEncoder()
        let actionData = try encoder.encode(action)
        guard let actionObject = try JSONSerialization.jsonObject(with: actionData) as? [String: Any] else {
            throw MppError.invalidJSON("session action did not encode to a JSON object")
        }
        let credential: [String: Any] = [
            "challenge": [
                "id": challenge.id,
                "realm": challenge.realm,
                "method": challenge.method,
                "intent": challenge.intent,
                "request": challenge.request,
            ].merging(optionalChallengeFields(challenge)) { current, _ in current },
            "payload": [
                "type": "session",
                "action": actionObject,
            ],
        ]
        let credentialData = try JSONSerialization.data(
            withJSONObject: credential, options: [.sortedKeys]
        )
        return "\(MppHeaders.paymentScheme) \(Base64URL.encode(credentialData))"
    }

    private static func optionalChallengeFields(_ challenge: PaymentChallenge) -> [String: Any] {
        var fields: [String: Any] = [:]
        if let expires = challenge.expires { fields["expires"] = expires }
        if let digest = challenge.digest { fields["digest"] = digest }
        if let opaque = challenge.opaque { fields["opaque"] = opaque }
        return fields
    }
}

// MARK: - Session consumer

/// Transport used by `SessionConsumer` to send commit payloads.
///
/// HTTP clients, queues, and in-process tests can all implement this.
/// The directive is passed alongside the payload so transports can use
/// `commitUrl`, `proof`, or other routing hints without repeating them in
/// the signed commit body.
public protocol CommitTransport: Sendable {
    func commit(directive: MeteringDirective, payload: CommitPayload) async throws -> CommitReceipt
}

/// Client-side consumer for session-metered deliveries.
public final class SessionConsumer<Transport: CommitTransport>: @unchecked Sendable {
    public let session: ActiveSession
    public let transport: Transport

    public init(session: ActiveSession, transport: Transport) {
        self.session = session
        self.transport = transport
    }

    /// Accept an envelope and return a delivery handle with `ack`/`commit`.
    public func accept<P: Codable & Sendable>(
        _ envelope: MeteredEnvelope<P>
    ) throws -> MeteredDelivery<Transport, P> {
        try validateDirective(envelope.metering)
        return MeteredDelivery(consumer: self, payload: envelope.payload, metering: envelope.metering)
    }

    /// Commit a directive directly, without constructing a delivery handle.
    @discardableResult
    public func commitDirective(_ directive: MeteringDirective) async throws -> CommitReceipt {
        try validateDirective(directive)
        let amount = try directive.amountBaseUnits()
        guard amount > 0 else {
            throw MppError.invalidTransaction("metered delivery amount must be greater than zero")
        }
        let voucher = try await session.prepareIncrement(amount)
        let payload = CommitPayload(deliveryId: directive.deliveryId, voucher: voucher)
        let receipt = try await transport.commit(directive: directive, payload: payload)
        try session.recordVoucher(payload.voucher)
        return receipt
    }

    private func validateDirective(_ directive: MeteringDirective) throws {
        let channelId = session.channelIdString()
        guard directive.sessionId == channelId else {
            throw MppError.invalidTransaction(
                "metered delivery session \(directive.sessionId) does not match active session \(channelId)"
            )
        }
    }
}

/// A delivered payload plus its metering directive. Call `ack` (or its
/// `commit` alias) after the application has processed `payload`.
public struct MeteredDelivery<Transport: CommitTransport, P: Codable & Sendable>: Sendable {
    private let consumer: SessionConsumer<Transport>
    public let payload: P
    public let metering: MeteringDirective

    init(consumer: SessionConsumer<Transport>, payload: P, metering: MeteringDirective) {
        self.consumer = consumer
        self.payload = payload
        self.metering = metering
    }

    @discardableResult
    public func ack() async throws -> CommitReceipt {
        try await consumer.commitDirective(metering)
    }

    @discardableResult
    public func commit() async throws -> CommitReceipt {
        try await ack()
    }
}
